subsys/bluetooth/host/adv.c

```diff
@@ -572,17 +581,26 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
 	while (!ad_stream_is_empty(&stream)) {
 		struct bt_hci_cp_le_set_ext_adv_data *set_data;
 		struct net_buf *buf;
+		size_t data_len, cmd_size;
 		int err;
 
-		buf = bt_hci_cmd_create(hci_op, sizeof(*set_data));
+		(void)memcpy(&backup, &stream, sizeof(stream));
+
+		data_len = ad_stream_read(&stream, NULL, sizeof(set_data->data));
+		cmd_size = data_len + offsetof(struct bt_hci_cp_le_set_ext_adv_data, data);
+
+		buf = bt_hci_cmd_create(hci_op, cmd_size);
 		if (!buf) {
 			return -ENOBUFS;
 		}
 
-		set_data = net_buf_add(buf, sizeof(*set_data));
+		set_data = net_buf_add(buf, cmd_size);
 
 		set_data->handle = adv->handle;
 		set_data->frag_pref = BT_HCI_LE_EXT_ADV_FRAG_ENABLED;
+
+		(void)memcpy(&stream, &backup, sizeof(stream));
+
```

Comment:
Hmm, so you are pretty much using `ad_stream_read` to get the data length, before creating the buffer, but in order to do that you had to refactor `ad_stream_read` and then also create a backup for the stream. 

In that case, I think a new function to simply get the data_len without having to complicate the `ad_stream_read` function as well as avoiding creating a backup would be a simpler and more clean solution.

Reply:
I also prefer the view of @alwa-nordic , because the current implementation is still the simplest, and rewriting one is not necessarily better.

Reply:
I like the idea that @alwa-nordic proposed, i.e. getting the total length in `ad_stream_new`. In that case, we should be able to avoid the save/restore functions, as well as the change to `ad_stream_read`, right?

Reply:
Please Review Again.